EDDI/EDDI.cs

```diff
@@ -1927,71 +1927,30 @@ private bool eventNearSurface(NearSurfaceEvent theEvent)
-                else
-                {
-                    int index = CurrentStarSystem.bodies.IndexOf(star);
-                    if (index != -1)
-                    {
-                        CurrentStarSystem.bodies[index] = theEvent.star;
-                    }
-                }
-                CurrentStarSystem.bodies = CurrentStarSystem.bodies.OrderBy(s => s.bodyId).ToList();
+            if (CurrentStarSystem == null) { return false; }
 
-                Logging.Debug("Saving data for scanned star " + theEvent.star.bodyname);
-                StarSystemSqLiteRepository.Instance.SaveStarSystem(CurrentStarSystem);
-            }
-            return CurrentStarSystem != null;
+            CurrentStarSystem.AddOrUpdateBody(theEvent.star);
+            StarSystemSqLiteRepository.Instance.SaveStarSystem(CurrentStarSystem);
+            return true;
         }
 
         private bool eventBodyScanned(BodyScannedEvent theEvent)
         {
             // We just scanned a body.  We can only proceed if we know our current star system
-            if (CurrentStarSystem != null)
-            {
-                Body body = CurrentStarSystem.bodies?.FirstOrDefault(b => b.bodyname == theEvent.bodyname);
-                if (body == null)
-                {
-                    Logging.Debug("Scanned body " + theEvent.bodyname + " is new - creating");
-                    CurrentStarSystem.bodies.Add(theEvent.body);
-                }
-                else
-                {
-                    if (body.scanned == null)
-                    {
-                        int index = CurrentStarSystem.bodies.IndexOf(body);
-                        if (index != -1)
-                        {
-                            // Update our body with the scan data.
-                            CurrentStarSystem.bodies[index] = theEvent.body;
-                        }
-                    }
-                    else
-                    {
-                        // We've already scanned this body. No need to repeat.
-                        return false;
-                    }
-                }
-                CurrentStarSystem.bodies = CurrentStarSystem.bodies.OrderBy(b => b.bodyId).ToList();
+            if (CurrentStarSystem == null) { return false; }
 
-                // Update the system reserve level, when appropriate
-                if (theEvent.body.reserveLevel != ReserveLevel.None)
-                {
-                    CurrentStarSystem.Reserve = theEvent.body.reserveLevel;
-                }
+            CurrentStarSystem.AddOrUpdateBody(theEvent.body);
 
-                Logging.Debug("Saving data for scanned body " + theEvent.bodyname);
-                StarSystemSqLiteRepository.Instance.SaveStarSystem(CurrentStarSystem);
+            // Update the system reserve level, when appropriate
+            if (theEvent.body.reserveLevel != ReserveLevel.None)
+            {
+                CurrentStarSystem.Reserve = theEvent.body.reserveLevel;
```

Comment:
We could automate this like we do system materials, so that it auto-updates as we add bodies.

Reply:
Good call.